src/query.rs

```diff
@@ -839,10 +749,100 @@ pub fn extract_smart_fields(
                 // if the requested position is available
                 let key = smt.alias.clone();
                 if smt.position - 1 < (found_vals[&smt.typed].len() as i32) {
-                    projection_values.insert(
-                        key,
-                        Some(found_vals[&smt.typed][(smt.position - 1) as usize].clone()),
-                    );
+                    let value = found_vals[&smt.typed][(smt.position - 1) as usize].clone();
+                    // match on subfield usage and validity of the subfield
+                    match (
+                        &smt.typed[..],
+                        &smt.subfield.as_ref().map_or(None, |m| Some(m.as_str())),
+                    ) {
+                        (SF_USER_AGENT, Some("name")) => {
+                            // TODO: Cache this parsing
+                            let parser = woothee::parser::Parser::new();
+                            match parser.parse(&value[..]) {
+                                Some(r) => {
+                                    projection_values.insert(key, Some(r.name.to_string()));
+                                }
+                                None => {
+                                    projection_values.insert(key, None);
+                                }
+                            }
+                        }
+                        (SF_USER_AGENT, Some("category")) => {
+                            // TODO: Cache this parsing
+                            let parser = woothee::parser::Parser::new();
+                            match parser.parse(&value[..]) {
+                                Some(r) => {
+                                    projection_values.insert(key, Some(r.category.to_string()));
+                                }
+                                None => {
+                                    projection_values.insert(key, None);
+                                }
+                            }
+                        }
+                        (SF_USER_AGENT, Some("browser_type")) => {
+                            // TODO: Cache this parsing
+                            let parser = woothee::parser::Parser::new();
+                            match parser.parse(&value[..]) {
+                                Some(r) => {
+                                    projection_values.insert(key, Some(r.browser_type.to_string()));
+                                }
+                                None => {
+                                    projection_values.insert(key, None);
+                                }
+                            }
+                        }
+                        (SF_USER_AGENT, Some("os")) => {
+                            // TODO: Cache this parsing
```

Comment:
What does this mean?

Reply:
We need to cache the `parser.parse` outcome since we may be re-parsing the line n times if `n` elements of `$user_agent` were requested.